crates/bevy_reflect/src/serde/de.rs

```diff
@@ -29,260 +47,273 @@ impl<'a, 'de> DeserializeSeed<'de> for ReflectDeserializer<'a> {
+                ));
+            }
+        };
+
+        match map.next_key::<&str>()? {
+            Some(type_fields::VALUE) => {
+                let registration = self.registry.get_with_name(type_name).ok_or_else(|| {
+                    de::Error::custom(format_args!("No registration found for {}", type_name))
+                })?;
+                let type_info = registration.type_info();
+                let value = map.next_value_seed(TypedReflectDeserializer {
+                    type_info,
+                    registry: self.registry,
+                })?;
+                Ok(value)
+            }
+            Some(type_fields::TYPE) => Err(de::Error::duplicate_field(type_fields::TYPE)),
+            Some(field) => Err(de::Error::unknown_field(field, &[type_fields::VALUE])),
+            None => Err(de::Error::invalid_length(
+                0,
+                &"two entries: `type` and `value`",
+            )),
+        }
     }
+}
 
-    fn visit_i16<E>(self, v: i16) -> Result<Self::Value, E>
-    where
-        E: de::Error,
-    {
-        Ok(Box::new(v))
-    }
+/// A deserializer for reflected types whose [`TypeInfo`] is known.
+///
+/// For non-value types, this will return the dynamic equivalent. For example, a
+/// deserialized struct will return a [`DynamicStruct`] and a `Vec` will return a
+/// [`DynamicList`].
+///
+/// [`TypeInfo`]: crate::TypeInfo
+/// [`DynamicStruct`]: crate::DynamicStruct
+/// [`DynamicList`]: crate::DynamicList
+pub struct TypedReflectDeserializer<'a> {
+    type_info: &'a TypeInfo,
+    registry: &'a TypeRegistry,
+}
 
-    fn visit_i32<E>(self, v: i32) -> Result<Self::Value, E>
-    where
-        E: de::Error,
-    {
-        Ok(Box::new(v))
-    }
+impl<'a, 'de> DeserializeSeed<'de> for TypedReflectDeserializer<'a> {
+    type Value = Box<dyn Reflect>;
 
-    fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
     where
-        E: de::Error,
+        D: serde::Deserializer<'de>,
     {
-        Ok(Box::new(v))
-    }
+        // Handle both Value case and types that have a custom `ReflectDeserialize`
+        let type_id = self.type_info.type_id();
+        let type_name = self.type_info.type_name();
+        let registration = self.registry.get(type_id).ok_or_else(|| {
+            de::Error::custom(format_args!("no registration found for {}", type_name))
+        })?;
+
+        if let Some(deserialize_reflect) = registration.data::<ReflectDeserialize>() {
+            let value = deserialize_reflect.deserialize(deserializer)?;
+            return Ok(value);
+        }
 
-    fn visit_f32<E>(self, v: f32) -> Result<Self::Value, E>
-    where
-        E: de::Error,
-    {
-        Ok(Box::new(v))
+        match self.type_info {
+            TypeInfo::Struct(struct_info) => {
+                let mut dynamic_struct = deserializer.deserialize_map(StructVisitor {
```

Comment:
Regarding `TypedReflectDeserializer`. I think it is ideal to use the specialized `deserialize_{tuple,tuple_stuct,etc.}` when matching `TypeInfo::{Tuple,TupleStruct,etc.}` because of format-specific optimizations. Using `deserlaize_seq` and `deserialize_map` in those circumstances will force the serialization format to serialize the plain text value of the field names etc. This might sound alright or unimportant for a plain-text format like `json` or `ron` (and even `msgpack`) but for optimized binary formats, the distinction is important, since it helps them:
- Elide field count info from the serialized format
- Elide field names as string as well

I'm writing my own deserializer. I'm going to try to do that and tell you how it looks like. Seems possible now that most strings in `TypeInfo` have `'static` lifetimes.

Reply:
Oooof, Why the hell does `deserialize_struct` take a `&'static [&'static str]`? Only `&[&'static str]` would be a bazilion times better :(

Reply:
Yep this is definitely the direction I think a lot of people would like to move towards. And yeah the whole `'static` requirement on everything makes it a bit trickier.

For now though, I'd consider such changes outside the scope of this PR. My goal here is to do something that's relatively uncontroversial (i.e. just reduce the number of "type" fields) so it can be approved/merged sooner.

Although I did enable custom serde formats (see description) as a kind of compromise towards that end goal in the meantime.

Reply:
This has now been addressed in a separate PR, #5723